pkg/util/acrtoken/acrtoken.go

```diff
@@ -0,0 +1,100 @@
+
+import (
+	"context"
+	"fmt"
+
+	azcontainerregistry "github.com/Azure/azure-sdk-for-go/services/containerregistry/mgmt/2019-06-01-preview/containerregistry"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/containerregistry"
+)
+
+type Manager interface {
+	CreateToken(context.Context, *api.RegistryProfile) error
+	CreatePassword(context.Context, *api.RegistryProfile) error
+	Delete(context.Context, *api.RegistryProfile) error
+}
+
+type manager struct {
+	env env.Interface
+
+	tokens     containerregistry.TokensClient
+	registries containerregistry.RegistriesClient
+
+	scopeMap string
+}
+
+func NewManager(env env.Interface, fpAuthorizer autorest.Authorizer) Manager {
+	return &manager{
+		env: env,
+
+		tokens:     containerregistry.NewTokensClient(env.SubscriptionID(), fpAuthorizer),
+		registries: containerregistry.NewRegistriesClient(env.SubscriptionID(), fpAuthorizer),
+
+		scopeMap: fmt.Sprintf("%s/scopeMaps/_repositories_pull", env.ACRResourceID()),
+	}
+}
+
+// Create create a token on our registry
+// see https://docs.microsoft.com/en-us/azure/container-registry/container-registry-repository-scoped-permissions
+func (m *manager) CreateToken(ctx context.Context, reg *api.RegistryProfile) error {
+	r, err := azure.ParseResourceID(m.env.ACRResourceID())
+	if err != nil {
+		return err
+	}
+
+	acrTokenName := "token-" + uuid.NewV4().String()
+	err = m.tokens.CreateAndWait(ctx, r.ResourceGroup, r.ResourceName, acrTokenName, azcontainerregistry.Token{
+		TokenProperties: &azcontainerregistry.TokenProperties{
+			ScopeMapID: &m.scopeMap,
+			Status:     azcontainerregistry.TokenStatusEnabled,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	reg.Name = r.ResourceName + ".azurecr.io"
+	reg.Username = acrTokenName
+	return nil
+}
+
+func (m *manager) CreatePassword(ctx context.Context, reg *api.RegistryProfile) error {
+	r, err := azure.ParseResourceID(m.env.ACRResourceID())
+	if err != nil {
+		return err
+	}
+	generateCredentialsParameters := azcontainerregistry.GenerateCredentialsParameters{
+		TokenID: to.StringPtr(fmt.Sprintf("%s/tokens/%s", m.env.ACRResourceID(), reg.Username)),
+		Name:    azcontainerregistry.TokenPasswordNamePassword1,
+	}
+	creds, err := m.registries.GenerateCredentials(ctx, r.ResourceGroup, r.ResourceName, generateCredentialsParameters)
+	if err != nil {
+		return err
+	}
+	if creds.Passwords == nil || len(*creds.Passwords) < 1 || (*creds.Passwords)[0].Value == nil {
+		return fmt.Errorf("generateCredentials returned empty passwords")
+	}
+	reg.Password = api.SecureString(*(*creds.Passwords)[0].Value)
+	return nil
+}
+
+func (m *manager) Delete(ctx context.Context, rp *api.RegistryProfile) error {
+	if rp.Name == "" || len(rp.Username) < 6 {
```

Comment:
This seems like an odd check.  Shouldn't you be checking `reg.Name == r.ResourceName + ".azurecr.io"` if you're going to?

Reply:
you need to make sure that the tokenName is not "" else it blocks for a very long time